File: backend/src/agents/state.ts
import { Annotation } from '@langchain/langgraph';
import { ToolResult } from '../tools/base';

export const AgentStateAnnotation = Annotation.Root({
  companyName: Annotation<string>(),
  llmProviderOverride: Annotation<string | null>({
    reducer: (oldVal, newVal) => newVal,
    default: () => null,
  }),
  ticker: Annotation<string | null>({
    reducer: (oldVal, newVal) => newVal,
    default: () => null,
  }),
  industry: Annotation<string | null>({
    reducer: (oldVal, newVal) => newVal,
    default: () => null,
  }),
  toolResults: Annotation<ToolResult[]>({
    reducer: (oldVal, newVal) => (oldVal || []).concat(newVal),
    default: () => [],
  }),
  currentStep: Annotation<string>({
    reducer: (oldVal, newVal) => newVal,
    default: () => '',
  }),
  completedSteps: Annotation<string[]>({
    reducer: (oldVal, newVal) => (oldVal || []).concat(newVal),
    default: () => [],
  }),
  structuredReport: Annotation<Record<string, any> | null>({
    reducer: (oldVal, newVal) => newVal,
    default: () => null,
  }),
  error: Annotation<string | null>({
    reducer: (oldVal, newVal) => newVal,
    default: () => null,
  }),
});

export type AgentState = typeof AgentStateAnnotation.State;

File: backend_python/app/agents/state.py
"""
Shared state object threaded through every node in the LangGraph pipeline.

Using a single TypedDict (rather than passing loose args between nodes) is
what makes LangGraph's graph wiring possible — every node reads from and
writes back into this same shape, and LangGraph handles merging.
"""
from typing import Annotated, TypedDict

from app.tools.base import ToolResult


def _last_value(_old: object, new: object) -> object:
    """Reducer: last node to write wins (default behavior, made explicit)."""
    return new


class AgentState(TypedDict, total=False):
    # --- input ---
    company_name: str
    llm_provider_override: str | None

    # --- resolved identity ---
    ticker: Annotated[str | None, _last_value]
    industry: Annotated[str | None, _last_value]

    # --- raw research, collected across nodes ---
    tool_results: Annotated[list[ToolResult], lambda old, new: (old or []) + new]

    # --- progress reporting, consumed by the streaming endpoint ---
    current_step: Annotated[str, _last_value]
    completed_steps: Annotated[list[str], lambda old, new: (old or []) + new]

    # --- final output ---
    structured_report: Annotated[dict | None, _last_value]
    error: Annotated[str | None, _last_value]
